File: expandabletext/src/main/java/com/kolyall/view/expandabletext/ExpandableView.java
package com.kolyall.view.expandabletext;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import lombok.Setter;
import lombok.experimental.Accessors;

/**
 * Created by Timofey Kovalenko on 22.01.2018.
 */
@Accessors(prefix = "m")
public class ExpandableView extends RelativeLayout {

    ExpandableTextView mExpandableTextView;
    ImageView mArrowView;

    Drawable mCollapseIcon;
    Drawable mExpandIcon;

    private int mTextColor;

    public ExpandableView(Context context) {
        super(context);
        setContentView(R.layout.view_expandable_view);
    }

    public ExpandableView(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.ExpandableView);
        mTextColor = typedArray.getColor(R.styleable.ExpandableView_android_textColor, ContextCompat.getColor(getContext(), R.color.colorAccent));
        mCollapseIcon = typedArray.getDrawable(R.styleable.ExpandableView_ev_collapseIcon);
        mExpandIcon = typedArray.getDrawable(R.styleable.ExpandableView_ev_expandIcon);
        typedArray.recycle();
        setContentView(R.layout.view_expandable_view);
    }

    protected void setContentView(int viewLayout) {
        inflate(getContext(), viewLayout, this);
        onViewCreated();
    }

    protected void onViewCreated() {
        setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        mExpandableTextView = findViewById(R.id.expandableTextView);
        mArrowView = findViewById(R.id.arrowView);
        mExpandableTextView.setTextColor(mTextColor);
        mArrowView.setImageDrawable(mExpandableTextView.isTrim() ? mExpandIcon : mCollapseIcon);
        mExpandableTextView.setOnExpandableTextViewListener(expanded -> {
            mArrowView.setImageDrawable(expanded ? mCollapseIcon : mExpandIcon);
            if (mOnExpandableTextViewListener != null) {
                mOnExpandableTextViewListener.onExpandChanged(expanded);
            }
        });
    }

    @Setter ExpandableTextView.OnExpandableTextViewListener mOnExpandableTextViewListener;

    public final void setText(CharSequence text) {
        mExpandableTextView.setText(text);
        mArrowView.setVisibility(mExpandableTextView.isModified() ? VISIBLE : GONE);
    }
}

File: expandabletext/src/main/java/com/kolyall/view/expandabletext/ExpandableView.kt
package com.kolyall.view.expandabletext

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import com.kolyall.view.expandabletext.ExpandableTextView.OnExpandableTextViewListener

class ExpandableView : RelativeLayout {
    private lateinit var mExpandableTextView: ExpandableTextView
    private lateinit var mArrowView: ImageView
    private var mCollapseIcon: Drawable? = null
    private var mExpandIcon: Drawable? = null
    private var mTextGravity = 0
    private var mTextColor = 0

    constructor(context: Context?) : super(context) {
        setContentView(R.layout.view_expandable_view)
    }

    constructor(
        context: Context,
        attrs: AttributeSet?
    ) : super(context, attrs) {
        val typedArray =
            context.obtainStyledAttributes(attrs, R.styleable.ExpandableView)
        mTextColor = typedArray.getColor(
            R.styleable.ExpandableView_android_textColor,
            ContextCompat.getColor(getContext(), R.color.colorAccent)
        )
        mCollapseIcon = typedArray.getDrawable(R.styleable.ExpandableView_ev_collapseIcon)
        mExpandIcon = typedArray.getDrawable(R.styleable.ExpandableView_ev_expandIcon)
        mTextGravity = typedArray.getInt(R.styleable.ExpandableView_ev_textGravity, -1)
        typedArray.recycle()
        setContentView(R.layout.view_expandable_view)
    }

    private fun setContentView(viewLayout: Int) {
        View.inflate(context, viewLayout, this)
        onViewCreated()
    }

    private fun onViewCreated() {
        layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        mExpandableTextView = findViewById(R.id.expandableTextView)
        mArrowView = findViewById(R.id.arrowView)
        mExpandableTextView.setTextColor(mTextColor)
        if (mTextGravity != -1) {
            setTextGravity(mTextGravity)
        }
        mArrowView.setImageDrawable(if (mExpandableTextView.isTrim) mExpandIcon else mCollapseIcon)
        mExpandableTextView.onExpandableTextViewListener =  object:OnExpandableTextViewListener{
            override fun onExpandChanged(expanded: Boolean) {
                mArrowView.setImageDrawable(if (expanded) mCollapseIcon else mExpandIcon)
                onExpandableTextViewListener?.onExpandChanged(expanded)
            }
        }
    }

    var onExpandableTextViewListener: OnExpandableTextViewListener? = null

    fun setText(text: CharSequence?) {
        mExpandableTextView.text = text
        mArrowView.visibility = if (mExpandableTextView.isModified) View.VISIBLE else View.GONE
    }

    fun setTextGravity(center: Int) {
        mExpandableTextView.gravity = center
    }
}